discoveryservice/module.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/storage"
+	"os"
+	"path"
+	"strings"
+	"time"
+)
+
+const ModuleName = "DiscoveryService"
+
+var ErrServerModeDisabled = errors.New("node is not a discovery server for this service")
+
+var _ core.Injectable = &Module{}
+var _ core.Runnable = &Module{}
+var _ core.Configurable = &Module{}
+var _ Server = &Module{}
+
+// var _ Client = &Module{}
+var retractionPresentationType = ssi.MustParseURI("RetractedVerifiablePresentation")
+
+func New(storageInstance storage.Engine) *Module {
+	return &Module{
+		storageInstance: storageInstance,
+	}
+}
+
+type Module struct {
+	config            Config
+	storageInstance   storage.Engine
+	store             *sqlStore
+	serverDefinitions map[string]Definition
+	services          map[string]Definition
+}
+
+func (m *Module) Configure(_ core.ServerConfig) error {
+	if m.config.Definitions.Directory == "" {
+		return nil
+	}
+	var err error
+	m.services, err = loadDefinitions(m.config.Definitions.Directory)
+	if err != nil {
+		return err
+	}
+	if len(m.config.Server.DefinitionIDs) > 0 {
+		// Get the definitions that are enabled for this server
+		serverDefinitions := make(map[string]Definition)
+		for _, definitionID := range m.config.Server.DefinitionIDs {
+			if definition, exists := m.services[definitionID]; !exists {
+				return fmt.Errorf("definition '%s' not found", definitionID)
+			} else {
+				serverDefinitions[definitionID] = definition
+			}
+		}
+		m.serverDefinitions = serverDefinitions
+	}
+	return nil
+}
+
+func (m *Module) Start() error {
+	var err error
+	m.store, err = newSQLStore(m.storageInstance.GetSQLDatabase(), m.services)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Module) Shutdown() error {
+	return nil
+}
+
+func (m *Module) Name() string {
+	return ModuleName
+}
+
+func (m *Module) Config() interface{} {
+	return &m.config
+}
+
+func (m *Module) Add(serviceID string, presentation vc.VerifiablePresentation) error {
+	definition, exists := m.services[serviceID]
+	if !exists {
+		return ErrServiceNotFound
+	}
+	if _, isMaintainer := m.serverDefinitions[serviceID]; !isMaintainer {
+		return ErrServerModeDisabled
+	}
+	if presentation.Format() != vc.JWTPresentationProofFormat {
+		return errors.New("only JWT presentations are supported")
+	}
+	// TODO: validate signature
+	if presentation.ID == nil {
+		return errors.New("presentation does not have an ID")
+	}
+	expiration := presentation.JWT().Expiration()
+	// VPs should not be valid for too short; unnecessary overhead at the server and clients.
+	// Also protects against expiration not being set at all. The factor is somewhat arbitrary.
+	minValidity := float64(definition.PresentationMaxValidity / 4)
+	if expiration.Sub(time.Now()).Seconds() < minValidity {
+		return fmt.Errorf("presentation is not valid for long enough (min %s)", time.Duration(minValidity)*time.Second)
+	}
```

Comment:
undocumented behaviour where the implementation is doing things different than the spec. I would just leave it out.

Reply:
or add minimum validity to the spec.
Once every minute won't trigger a rate limit but is still way too much